### DISCUSSION TITLE
Issues with fields having similar relative xPaths.

### DISCUSSION THREAD

**Discussion by pankajgulia111** [OP-TED/eForms-SDK#445] [Q&A] [answered]
Hi all,
i have  issue with handling of the fields "BT-743-Lot", "BT-764-Lot", and "BT-736-Lot", which have the following relative XPath: `"cac:ContractExecutionRequirement/cbc:ExecutionRequirementCode[@listName='einvoicing']"`, `"cac:ContractExecutionRequirement/cbc:ExecutionRequirementCode[@listName='ecatalog-submission']"`, and `"cac:ContractExecutionRequirement/cbc:ExecutionRequirementCode[@listName='reserved-execution']"`. These fields have the attribute "repeatable.value" set to "false" in "fields.json".
The approach i am taking is dividing the xpath into two `cac:ContractExecutionRequirement` and `cbc:ExecutionRequirementCode[@listName='reserved-execution']` parts. First checking if there is an xml element by tag `cbc:ExecutionRequirementCode`if yes then treating this tag as parent i assign `cbc:ExecutionRequirementCode[@listName='abc']`
as its child.
In "**cn_24_minimal.xml**", we can see that the `cac:TenderingTerms` element contains `cac:ContractExecutionRequirement` three times. However, as these fields have `"cac:ContractExecutionRequirement"`in their relative XPath, this results in creating one node of `"cac:ContractExecutionRequirement"` with all three fields as its child, which makes the XML file invalid.
Do we treat these fields repeatable?

**VALID XML**
```
<cac:ContractExecutionRequirement>
            <cbc:ExecutionRequirementCode listName="einvoicing">required</cbc:ExecutionRequirementCode>
         </cac:ContractExecutionRequirement>
         <cac:ContractExecutionRequirement>
            <cbc:ExecutionRequirementCode listName="reserved-execution">no</cbc:ExecutionRequirementCode>
         </cac:ContractExecutionRequirement>
         <cac:ContractExecutionRequirement>
            <cbc:ExecutionRequirementCode listName="ecatalog-submission">allowed</cbc:ExecutionRequirementCode>
         </cac:ContractExecutionRequirement>
```
         
**INVALID XML**
``` 
<cac:ContractExecutionRequirement>
            <cbc:ExecutionRequirementCode listName="einvoicing">required</cbc:ExecutionRequirementCode>
            <cbc:ExecutionRequirementCode listName="reserved-execution">no</cbc:ExecutionRequirementCode>
            <cbc:ExecutionRequirementCode listName="ecatalog-submission">allowed</cbc:ExecutionRequirementCode> </cac:ContractExecutionRequirement>
```

Thanks
Pankaj

**Comment by rouschr** [comment]
Hi,
how did you generate the XML and which SDK are you using?

**Reply by pankajgulia111** [reply]
Hi,
I am using SDK 1.6.
I modified the eforms-notice-editor example , the flow is visual Model to conceptual model and generating XML from conceptual model.

**Comment by elerik** [comment]
this is quite tricky. I used whole xPathRelative as identificator of "forks" where i create whole new node child parallel to any other existing childs within parent node. And well they are technically not repeatable. They are simply diference with just different attributes. Thats it. 

There are just few situations of this. 

There is my list of patterns i identity during creating XML

cbc:Xyz -> field value is element value
cbc:Xyz/@listName -> field value is elements attribute value named listName (or whatever) (for example BT-26(a)-Lot)
cbc:Xyz/cbc:Yxz -> field value is node Xyz and element value of Yxz (if some Xyz already exists create new node and continue adding values in that node)
cbc:Xyz[@listName='attributeValue'] -> field value is element value and append attribute listName="attributeValue"
cbc:Xyz/cbc:Yxz[not(cac:Zyx)] -> not solved yet :D (for example BT-71-Lot) 

and so far this is what created me valid XML (CVS answer success) which is possible to submit via API.

**Comment by rouschr** [comment, marked as answer]
I had a look at this and after checking with a colleague I also think that for field relative xpath it should not try to reuse existing elements, it should just recreate the entire relative path.
By doing this as you said you would end up with valid XML.
I will later update the editor demo project to reflect this change, you can see it applied in this branch:
https://github.com/OP-TED/eforms-notice-editor/tree/feature/fix-xsd-path-loading
